### PR TITLE
Support contract mapping

### PR DESCRIPTION
**Pull request by nolag**

### CODE REVIEW

File: core/services/relay/evm/chain_reader.go [hunk 1]
```diff
@@ -29,17 +29,17 @@ type ChainReaderService interface {
...
-	lp         logpoller.LogPoller
-	codec      commontypes.RemoteCodec
-	client     evmclient.Client
-	contractID common.Address
-	events     map[string]common.Hash
+	lggr     logger.Logger
+	lp       logpoller.LogPoller
+	codec    commontypes.RemoteCodec
+	client   evmclient.Client
+	bindings map[string]map[string]common.Address
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:36]: add a comment here that indicates what key value means in these maps

File: core/services/relay/evm/chain_reader.go [hunk 2]
```diff
@@ -29,17 +29,17 @@ type ChainReaderService interface {
...
+	events   map[string]map[string]common.Hash
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:37]: same comment as for bindings

File: core/services/relay/evm/chain_reader.go [hunk 3]
```diff
@@ -169,53 +183,55 @@ func addEncoderDef(name string, method abi.Method, parsed *parsedTypes, chainRea
...
 			switch chainReaderDefinition.ReadType {
 			case types.Method:
-				err = addMethods(chainReadingDefinitionName, contractAbi, chainReaderDefinition, parsed)
+				err = addMethods(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
 			case types.Event:
 				var hash common.Hash
-				hash, err = addEventTypes(chainReadingDefinitionName, contractAbi, chainReaderDefinition, parsed)
-				events[chainReadingDefinitionName] = hash
+				hash, err = addEventTypes(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
+				eventMap[typeName] = hash
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:218]: eventMap[typeName], err = addEventTypes(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
```suggestion
				hash, err = addEventTypes(contractName, typeName, contractAbi, chainReaderDefinition, parsed)
				eventMap[typeName] = hash
```

File: core/services/relay/evm/chain_reader.go [hunk 4]
```diff
@@ -104,59 +106,71 @@ func (cr *chainReader) getLatestValueFromContract(ctx context.Context, bc common
...
 		return fmt.Errorf("method: %q doesn't exist", chainReaderDefinition.ChainSpecificName)
 	}
 
-	if err := addEncoderDef(name, method, parsed, chainReaderDefinition); err != nil {
+	if err := addEncoderDef(contractName, typeName, method, parsed, chainReaderDefinition); err != nil {
 		return err
 	}
 
-	return addDecoderDef(name, method.Outputs, parsed, chainReaderDefinition)
+	return addDecoderDef(contractName, typeName, method.Outputs, parsed, chainReaderDefinition)
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:170]: naming seems a bit inconsistent.  `addDecoderDef`  should have its `methodName` param rename to typeName as it is used both when adding methods and events, and `addEncoderDef` should should have its `typeName` param renamed to  `methodName` as it is only used by `addMethods`.
Renaming both to be `typeName` also makes sense

File: core/services/relay/evm/chain_reader.go [hunk 5]
```diff
@@ -104,59 +106,71 @@ func (cr *chainReader) getLatestValueFromContract(ctx context.Context, bc common
...
 	if !methodExists {
 		return common.Hash{}, fmt.Errorf("method: %s doesn't exist", chainReaderDefinition.ChainSpecificName)
 	}
 
-	return event.ID, addDecoderDef(name, event.Inputs, parsed, chainReaderDefinition)
+	return event.ID, addDecoderDef(contractName, typeName, event.Inputs, parsed, chainReaderDefinition)
 }
 
-func addMethods(name string, abi abi.ABI, chainReaderDefinition types.ChainReaderDefinition, parsed *parsedTypes) error {
+func addMethods(contractName, typeName string, abi abi.ABI, chainReaderDefinition types.ChainReaderDefinition, parsed *parsedTypes) error {
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:160]: I like name or methodName better than typeName here, same for `addEvents`

Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:160]: Not really related to this PR, but just realised that these should be renamed to `addEventType` and add `addMethodType`. And maybe add comment that explains that by add we mean add to codec.

**Review by ilija42 — CHANGES_REQUESTED** [review]
Mostly code nits

File: core/services/relay/evm/chain_reader.go (hunk 4 above)
Comment (nolag) [reply core/services/relay/evm/chain_reader.go:178]: contractName, methodName makes sense.  typeName will actually be the combination that's created in wrap.  I'll. give the file a quick read-over.

**Review by ilija42 — APPROVED** [review]
Just a note, its a bit hard to see the changes when its force pushed into a single commit. Maybe do this after final approval, so its easier to re-review?

**Pull request merged**